Message: support body to be a callback

petrkotek opened the pull request:
## What?
In `Sabre\HTTP\MessageInterface` (and `Sabre\HTTP\Message`), support `$body` to be a callback outputting the body into `php://output`

## Why?
Some responses may be big, so it's not scalable to hold the whole response in a variable (in memory), see https://github.com/fruux/sabre-dav/pull/890.

## Limitation
`getBodyAsString()` and `getBodyAsStream()` doesn't currently work with the callback (throws an exception). It's implementation seem to be a bit complicated - mostly because we are able to produce output only once, but i'm open to ideas.

## Usage example from sabre/dav
Snippet from `sabre/dav` to make `generateMultiStatus()` method work as a callback:
```php
    function generateMultiStatus($fileProperties, $strip404s = false) {
        
        $w = $this->xml->getWriter();
        return function() use ($fileProperties, $strip404s, $w) {
            $w->openUri('php://output');
            $w->contextUri = $this->baseUri;
            $w->startDocument();
            $w->startElement('{DAV:}multistatus');

            foreach ($fileProperties as $entry) {
                $href = $entry['href'];
                unset($entry['href']);
                if ($strip404s) {
                    unset($entry[404]);
                }
                $response = new Xml\Element\Response(
                    ltrim($href, '/'),
                    $entry
                );
                $w->write([
                    'name'  => '{DAV:}response',
                    'value' => $response
                ]);
                // maybe we should flush() from time to time here as well
            }

            $w->endElement();
            $w->endDocument();
            $w->flush();
        };

    }
```

staabm commented on lib/Sapi.php:
```diff
@@ -71,6 +71,11 @@ static function sendResponse(ResponseInterface $response) {
         if (is_null($body)) return;
 
+        if (is_callable($body)) {
+            $body();
```
shouldn't we pass the $response object into the callback so it can handle the case of existing Content-Length header etc? (like wie do for non-callback bodies a few lines below)

petrkotek replied:
i'm not sure if it's necessary - i think that the same layer which is setting the headers (e.g. content length) is also setting the body (callback), so it can also pass the content length to the callback (e.g. `function() use ($contentLength) {...}`

staabm replied:
looking at your generateMultiStatus example tells me that this assumption doesnt work?

evert replied:
I'm with @petrkotek on this. `generateMultiStatus` really should move out of `Sabre\DAV\Server` anyway. It's a leftover from the extremely early days.

Usually I think it's reasonable to assume that the thing that sets the body, also has the ability to access the response object.

petrkotek commented:
ping @evert / @staabm: will you have a chance to look into this? thanks!

petrkotek commented:
<bump>

can i get some 👀 on this? thanks!

evert commented on lib/Message.php:
```diff
@@ -74,6 +77,9 @@ function getBodyAsString() {
             return '';
         }
+        if (is_callable($body)) {
+            throw new \UnexpectedValueException('Callback to string not supported');
```
We can definitely support this with output buffering.

evert replied:
`ob_start` / `ob_get_clean` and stuff ;)

evert commented on lib/Message.php:
```diff
@@ -53,6 +53,9 @@ function getBodyAsStream() {
             return $stream;
         }
+        if (is_callable($body)) {
+            throw new \UnexpectedValueException('Callback to stream not supported');
```
This can also definitely be done, and would be nice to have! With output buffering this can be turned into a stream.

evert replied:
More than 'nice to have' actually ;) I think this needs to be an absolute requirement to not break existing stuff using the HTTP response.

petrkotek commented:
so in the last commit i implemented `getBodyAsString()` and `getBodyAsStream()` for the case when body is a callback. I'm not sure if there is more effective way how to convert the `php://output` into stream -- at the moment there is an intermediate string representation used

I also removed the commit 47c4b1f which was changing `$body()` to `$body($response)`

staabm commented on lib/Message.php:
```diff
@@ -312,4 +320,23 @@ function getHttpVersion() {
+            throw new \RuntimeException('Cannot start output buffering');
+        }
+        $callback();
+        $content = ob_get_contents();
```
`ob_get_clean()` would do the same in 1 step

petrkotek replied:
yay, very good point. i don't use this `ob_*` stuff too much :) updated

(note that i can always rebase & squash the commits if needed)

evert commented on lib/Message.php:
```diff
@@ -318,4 +320,23 @@ function getHttpVersion() {
+    private function captureCallbackOutput($callback)
+    {
+        $success = ob_start();
+        if ($success === false) {
```
You don't really need this check. I've never encountered a situation where this can possibly fail.

petrkotek replied:
✅

evert commented on tests/HTTP/MessageTest.php:
```diff
@@ -258,11 +256,15 @@ function testHasHeaders() {
     {
-        return function() {
+        return function() use ($content) {
             $fd = fopen('php://output', 'r+');
```
This is a nitpick and not that important, but instead of opening a stream to `php://output` and writing the contents there, you can just do:

`echo $content;`.

And this will have an identical effect, and possibly even a bit faster. Not important for unittests though.

petrkotek replied:
thanks, good suggestions; fixed ✅
